### PR TITLE
feature: add nextcloud integration

### PR DESCRIPTION
**Pull request by strachanbrad**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/PhotoboothProject/photobooth/blob/dev/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "x" next to an item)
- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Other, please explain:

#### What changes did you make? (Give an overview)
Changed qrcode.php to remove the link generation.
Created sharelink.php to use Nextclooud Share API to generate share links and retrieve them to be sent to qrcode.php.
Created nc_copy_on_mnt.sh script to automatically copy photos to a Nextcloud mount.
Changed api.showQr in core.js to call sharelink.php then send the link to qrcode.php
Added a config section under expert to Nextcloud related settings. Related peripherals were changed as-well
Changed services_start and services_stop to account for nc_copy_on_mnt
Added and FAQ section to explain the feature and addition packages to install/ system configurations

everything is set to default to the original functionality of qrcode.php if Nextcloud is disabled or there is mount/network issue. 

#### Is there anything you'd like reviewers to focus on?
This is my first contribution to a project ever so be critical please. I'm not even sure if this should be in the scope of the project as there is a little bit of system configuration to be done. I want to make sure that the services_start and services_stop works as intended, and that nc_copy_on_mnt does not use up too much system resources. I also have a pretty internet connection for the upload of photos and the preceding generation of the Nextcloud share is pretty quick for me, I'd like to make sure that the lag to get the qrcode isn't too cumbersome.

Also I could not figure Crowdin sorry.

### CODE REVIEW

**Comment by andi34** [comment]
Hey and thanks for your contribution!
I am just rewriting the almost the whole PHP parts of the project (https://github.com/PhotoboothProject/photobooth/pull/200). There's some small things left i need/like to adjust before i can merge.
Your changes to the API files definitely conflict with it, which is why I am going to review your changes deeper after that.
Some of your functions from the sharelink API endpoints should then go into the Helper or Photobooth class. Also some error handling should be added.
With "yarn format" your changes should be formatted to the desired text/code formatting.

Overall it's a good idea and a functionality requested by users in the past already.

Again, thanks a lot for your contribution!

File: api/qrcode.php [hunk 1]
```diff
@@ -2,31 +2,32 @@
...
 
-$filename = $_GET['filename'];
+$filename = $_GET['share_link'];
 
 if ($config['qr']['append_filename']) {
-    $url = $config['qr']['url'] . $filename;
+	$url = $filename;
 } else {
-    $url = $config['qr']['url'];
+	$url = $config['qr']['url'] . "/s/qyGko4KNoZsJY39";
```
Comment (andi34) [inline api/qrcode.php]: What's the `/s/qyGko4KNoZsJY39` for?

**Comment by andi34** [comment]
Some questions already came in my mind:
Any changes needed for the QR inside the gallery?
Any changes needed for printing the QR onto the image?

If yes, should be looked at after the big API changes (or if you like rebase your current changes onto that branch).

File: api/qrcode.php (hunk 1 above)
Comment (strachanbrad) [reply api/qrcode.php]: Apologies that was missed from early testing. It was just a fall back share url from when this was just for me. It all has been moved to sharlink.php

**Comment by strachanbrad** [comment]
No changes in the gallery required, API.showQr is used in both cases (after an image is taken, and in the gallery)

**Comment by strachanbrad** [comment]
It sure about the QR printing onto an image, I'll take a look at that

**Comment by andi34** [comment]
> It sure about the QR printing onto an image, I'll take a look at that

Don't put to much energy on it for the current sources, if you look at it make sure to base it onto https://github.com/PhotoboothProject/photobooth/pull/200

**Comment by andi34** [comment]
`$config['nextcloud']['share'];` is a static URL? If yes, we should remove that option as the general qr url can be adjusted as needed and used instead

**Comment by strachanbrad** [comment]
> `$config['nextcloud']['share'];` is a static URL? If yes, we should remove that option as the general qr url can be adjusted as needed and used instead
> 
> 

Yes it is. And it's separate from the Qr url because, generating the share link of an image may take time in some cases, therefore it can fall back to the share url for the nextcloud folder instead of the specific image. If there is a nextwork issue or a mount issue it will fall back to the qr url. 

Maybe there is a better way of doing this, or like you said just remove it.

**Comment by andi34** [comment]
Thanks for the explanation. I think the qr url should be the fallback as network issues won't harm if it's static, appending the Filename is optional and can be disabled.

I'll work something out you can test, I'll let you know once it's ready for testing :)

**Comment by andi34** [comment]
I've added your changes here and pushed some updates to the code

https://github.com/PhotoboothProject/photobooth/commits/wip

Not tested and not finished yet, just to let you know

**Comment by strachanbrad** [comment]
> I've added your changes here and pushed some updates to the code
> 
> 
> 
> https://github.com/PhotoboothProject/photobooth/commits/wip
> 
> 
> 
> Not tested and not finished yet, just to let you know

Thank you for your update, and thank you for efforts and work!!

**Comment by andi34** [comment]
Adjusted some things and fixed some small errors. If you like to test:
```
sudo -u www-data -s
cd /var/www/html
git remote add pb https://github.com/PhotoboothProject/photobooth.git
git fetch pb
git checkout pb/wip
git submodule update --init
yarn install
yarn build
exit
```

**Comment by strachanbrad** [comment]
> Adjusted some things and fixed some small errors. If you like to test:
> 
> ```
> 
> sudo -u www-data -s
> 
> cd /var/www/html
> 
> git remote add pb https://github.com/PhotoboothProject/photobooth.git
> 
> git fetch pb
> 
> git checkout pb/wip
> 
> git submodule update --init
> 
> yarn install
> 
> yarn build
> 
> exit
> 
> ```

Ok I've embarked on testing, I know it's a work in progress for you. How would you like me to go about submitting fixes?

Also I have found out a way to send the images to nextcloud via curl in php, I think it would simplify the whole process to the user as they can just set their config with no need to mess arround with services or install additional packages. And it will work on windows.

I personally like having the WebDAV mounted directory as it's just one persistent connection. For my use case at events that's desirable, however I can see the barrier to entry. Your thoughts would be appreciated, I can make another pull request base on wip so you can take a look when it's done.

I was thinking I could move the nextcloud config to advanced, but have an option in expert if the user chooses to go with a mounted solution? Your input will guide my direction.

Cheers I know your a busy person.

**Comment by strachanbrad** [comment]
FYI if you need a hand with anything outside this pull request let me know I can make some time. I can write documentation, report/fix bugs, feature requests on the php/js side (not so much nodejs, I've dabbled).

**Comment by andi34** [comment]
Hey and thanks a lot!
If you're ok you can grab a copy of the wip Branch and continue to work on that. I can pull those changes to the wip branch and later we're going to update your PR to be based on the source of the dev branch if you're ok with that.

The wip Branch will change, but don't worry about that.

You can show/hide the options as needed, if there's an option which should be in expert or experimental view only it's fine. Options can also depends on the server OS (https://github.com/PhotoboothProject/photobooth/commit/3129cb3936035cefda73333c7577b37e213c5b75).

The api/lib rework has an own PR, if you find any bugs or have improvements you can create a PR for that branch too.

I am happy for every helping hand, my skills are self-learned and I am sure there's air for improvements 😊😉

Working with classes should give the way to write PHP Unit tests later, something I need to look at in future.

**Comment by strachanbrad** [comment]
> Hey and thanks a lot! If you're ok you can grab a copy of the wip Branch and continue to work on that. I can pull those changes to the wip branch and later we're going to update your PR to be based on the source of the dev branch if you're ok with that.
> 
> The wip Branch will change, but don't worry about that.
> 
> You can show/hide the options as needed, if there's an option which should be in expert or experimental view only it's fine. Options can also depends on the server OS ([3129cb3](https://github.com/PhotoboothProject/photobooth/commit/3129cb3936035cefda73333c7577b37e213c5b75)).
> 
> The api/lib rework has an own PR, if you find any bugs or have improvements you can create a PR for that branch too.
> 
> I am happy for every helping hand, my skills are self-learned and I am sure there's air for improvements blushwink
> 
> Working with classes should give the way to write PHP Unit tests later, something I need to look at in future.

Roger that!

I'm the same way self taught, I think alot of us are.

**Comment by andi34** [comment]
And if you like you can join us at telegram, conversations might be easier there (don't be scared about the German talk, most active user seem to be from Germany like me)

File: admin/components/footer.admin.php [hunk 2]
```diff
@@ -1,7 +1,7 @@
     <div class="pageLoader w-full h-full fixed top-0 left-0 z-50 hidden place-items-center [&.isActive]:grid">
         <div class="w-full h-full left-0 top-0 z-10 absolute bg-black bg-opacity-60"></div>
         <div class="px-4 py-6 rounded-md bg-white shadow-md flex flex-col items-center justify-center relative z-20 text-center">
-            <?php echo getLoader("sm"); ?>
+            <?php echo getLoader("lg"); ?>
```
Comment (andi34) [inline admin/components/footer.admin.php]: Needed for Nextcloud? Seperate?

**Comment by andi34** [comment]
Once https://github.com/PhotoboothProject/photobooth/pull/231 merged we need to take care about

**Review by andi34 — CHANGES_REQUESTED** [review]
Needs rebase.
QR on result screen and video qr need to be fixed.

**Comment by andi34** [comment]
Needs to be rewritten if still interesting, upload via scp from post-photo command should be fine too.
https://photoboothproject.github.io/FAQ#how-to-upload-pictures-to-a-remote-server-after-picture-has-been-taken

**Pull request closed without merging**